File: 0215-kth-largest-element-in-an-array/0215-kth-largest-element-in-an-array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        min_heap = []

        for num in nums:
            heapq.heappush(min_heap, num)
            # Pop once heap size reaches k - N * LogK(Time)
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        return min_heap[0]

File: 0215-kth-largest-element-in-an-array/0215-kth-largest-element-in-an-array.java
// MinHeap: Time: O(Nlogk); Space: O(k)
class Solution {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for(int num:nums){
            heap.offer(num);

            if (heap.size() > k){
                heap.poll();
            }
        }

        return heap.peek();
    }
}

// MaxHeap: Time: O(N) + O(klogN); Space: O(N)
// class Solution {
//     public int findKthLargest(int[] nums, int k) {
//         PriorityQueue<Integer> max_heap = new PriorityQueue<>(Collections.reverseOrder());
//         for(int num:nums){
//             max_heap.offer(num);
//         }

//         for(int i=0; i < k-1; i++){
//             max_heap.poll();
//         }

//         return max_heap.peek();
//     }
// }